packages/dev/core/src/Misc/snapshotRenderingHelper.ts

```diff
@@ -0,0 +1,228 @@
+        this._scene = scene;
+        this._engine = scene.getEngine();
+
+        if (!this._engine.isWebGPU) {
+            return;
+        }
+
+        this._options = {
+            morphTargetsNumMaxInfluences: 20,
+            ...options,
+        };
+
+        this._engine.snapshotRenderingMode = Constants.SNAPSHOTRENDERING_FAST;
+
+        this.fixMeshes();
+
+        this._onResizeObserver = this._engine.onResizeObservable.add(() => {
+            const save = this._engine.snapshotRendering;
+            this._engine.snapshotRendering = false;
+            if (save) {
+                this.enableSnapshotRendering();
+            }
+        });
+
+        this._scene.onBeforeRenderObservable.add(() => {
+            if (!this._engine.snapshotRendering || this._engine.snapshotRenderingMode !== Constants.SNAPSHOTRENDERING_FAST) {
+                return;
+            }
+
+            // Animate skeletons
+            scene.skeletons.forEach((skeleton) => skeleton.prepare(true));
+
+            for (const mesh of scene.meshes) {
+                if (mesh.infiniteDistance) {
+                    mesh.transferToEffect(mesh.computeWorldMatrix(true));
+                }
+
+                if (mesh.skeleton) {
+                    mesh.transferToEffect(mesh.computeWorldMatrix(true));
+                }
+
+                if (mesh.morphTargetManager && mesh.subMeshes) {
+                    // Make sure morph target animations work
+                    for (const subMesh of mesh.subMeshes) {
+                        const dw = subMesh._drawWrapper;
+                        const effect = dw.effect;
+                        if (effect) {
+                            const dataBuffer = (dw.drawContext as WebGPUDrawContext).buffers[WebGPUShaderProcessor.LeftOvertUBOName];
+                            const ubLeftOver = (effect._pipelineContext as WebGPUPipelineContext)?.uniformBuffer;
+                            if (dataBuffer && ubLeftOver && ubLeftOver.setDataBuffer(dataBuffer)) {
+                                mesh.morphTargetManager._bind(effect);
+                                BindMorphTargetParameters(mesh, effect);
+                                ubLeftOver.update();
+                            }
+                        }
+                    }
+                }
+            }
+        });
+    }
+
+    /**
+     * Enable snapshot rendering
+     * Use this method instead of engine.snapshotRendering=true, to make sure everything is ready before enabling snapshot rendering.
+     */
+    public enableSnapshotRendering() {
+        if (!this._engine.isWebGPU) {
+            return;
+        }
+
+        this._scene.executeWhenReady(() => {
+            // Make sure a full frame is rendered before enabling snapshot rendering, so use "+2" instead of "+1"
+            this._executeAtFrame(this._engine.frameId + 2, () => {
+                this._engine.snapshotRendering = true;
+            });
+        });

```

Comment:
This function is async but doesn't expose the asynchrony in any way. This means a caller could call disableSnapshotRendering after enableSnapshotRendering and you'd get the wrong behavior. For the viewer, I will effectively have to ref count this somewhere (e.g. when you have overlapping async operations for creating the skybox and loading a model). We could ref count in this helper, or we could just allow for cancellation and leave it to the caller to manage ref counting if they need it (e.g. the viewer).

```suggestion
    public enableSnapshotRendering(abortSignal: AbortSignal) {
        if (!this._engine.isWebGPU) {
            return;
        }

        this._scene.executeWhenReady(() => {
            // Make sure a full frame is rendered before enabling snapshot rendering, so use "+2" instead of "+1"
            this._executeAtFrame(this._engine.frameId + 2, () => {
                if (!abortSignal.aborted) {
                    this._engine.snapshotRendering = true;
                }
            });
        });

```

Reply:
I preferred to handle the problem internally. Please let me know if this works for you.

Reply:
Understood, I updated the implementation. Let me know if it can work for you.